subprojects/core/src/main/java/org/gradle/process/internal/DefaultJavaExecSpec.java

```diff
@@ -38,45 +38,36 @@
 
 public abstract class DefaultJavaExecSpec extends DefaultJavaForkOptions implements JavaExecSpec {
 
-    private final ProcessArgumentsSpec argumentsSpec = new ProcessArgumentsSpec(new ProcessArgumentsSpec.HasExecutable() {
-        @Override
-        public String getExecutable() {
-            return DefaultJavaExecSpec.this.getExecutable().get();
-        }
-
-        @Override
-        public void setExecutable(Object executable) {
-            DefaultJavaExecSpec.this.executable(executable);
-        }
-    });
-
     private final Property<String> mainClass;
     private final Property<String> mainModule;
     private final ModularitySpec modularity;
     private final ListProperty<String> jvmArguments;
 
-    private final FileCollectionFactory fileCollectionFactory;
-    private ConfigurableFileCollection classpath;
+    /**
+     * In scopes JavaModuleDetector is not available, e.g. in Process isolation Worker. In such cases, this field will be null.
+     */
+    @Nullable
+    private final JavaModuleDetector javaModuleDetector;
 
     @Inject
     public DefaultJavaExecSpec(
         ObjectFactory objectFactory,
         PathToFileResolver resolver,
-        FileCollectionFactory fileCollectionFactory
+        FileCollectionFactory fileCollectionFactory,
+        Factory<JavaModuleDetector> javaModuleDetector
     ) {
         super(objectFactory, resolver, fileCollectionFactory);
         this.jvmArguments = objectFactory.listProperty(String.class);
         this.mainClass = objectFactory.property(String.class);
         this.mainModule = objectFactory.property(String.class);
         this.modularity = objectFactory.newInstance(DefaultModularitySpec.class);
-        this.fileCollectionFactory = fileCollectionFactory;
-        this.classpath = fileCollectionFactory.configurableFiles("classpath");
+        this.javaModuleDetector = javaModuleDetector.create();
```

Comment:
❓ Is there a reason to pass in a factory instead of the instance itself?

Reply:
The reason is that in some cases, for example [here](https://github.com/gradle/gradle/blob/d2d46b78f6971c89f8f8ce9352ff2cacf8be9adf/subprojects/core/src/main/java/org/gradle/process/internal/DefaultExecActionFactory.java#L167), JavaModuleDetector can be `null`, but we cannot inject null. So we inject nonnull `Factory`